packages/catlog/src/stdlib/analyses/reachability/mod.rs

```diff
@@ -0,0 +1,222 @@
+//! Reachability analyses of models.
+
+use itertools::Itertools;
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use crate::dbl::modal::model::{ModalDblModel, ModalOb};
+use crate::zero::QualifiedName;
+use crate::zero::{FinSet, MutMapping};
+use std::collections::HashMap;
+#[cfg(feature = "serde-wasm")]
+use tsify::Tsify;
+
+/// Data defining a mass-action ODE problem for a model.
+#[derive(Clone)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde-wasm", derive(Tsify))]
+#[cfg_attr(
+    feature = "serde-wasm",
+    tsify(into_wasm_abi, from_wasm_abi, hashmap_as_object)
+)]
+pub struct ReachabilityProblemData {
+    /// Map from morphism IDs to rate coefficients (nonnegative reals).
+    pub tokens: HashMap<QualifiedName, i32>,
+
+    /// Map from object IDs to initial values (nonnegative reals).
+    pub forbidden: HashMap<QualifiedName, i32>,
+}
+
+/// The "Region Algebra for Petri Nets" algorithm from Ch 31 of
+/// the Handbook of Model Checking: "Symbolic Model Checking in Non Boolean
+/// Domains".
+pub fn reachability(m: &ModalDblModel, data: ReachabilityProblemData) -> Result<bool, String> {
+    // Convert model into a pair of matrices
+    //--------------------------------------
+
+    // Get a canonical ordering of the objects
+    let ob_vec: Vec<QualifiedName> = data.tokens.keys().cloned().sorted().collect();
+    let ob_inv: HashMap<QualifiedName, usize> =
+        ob_vec.iter().enumerate().map(|(x, y)| (y.clone(), x)).collect();
+    let n_p: usize = ob_vec.len();
+    if n_p == 0 {
+        return Ok(true);
+    }
+
+    // Get a canonical ordering of the homs
+    let hom_vec: Vec<QualifiedName> = m.mor_generators.edge_set.iter().sorted().collect();
```

Comment:
To avoid the private field access, the RHS can be replaced with `m.mor_generators().sorted().collect()`